File: findMaxDistance.cpp
#include<bits/stdc++.h>
using namespace std;

pair<int,int>bfs(int node, int n, vector<int>adj[]){
        queue<pair<int,int>>q;
        q.push({node, 0});
 
        vector<bool>vis(n+1, false);
 
        vis[node] = true;
 
        pair<int,int>lastVis;
 
        while(!q.empty()){
            
              lastVis = q.front();
              q.pop();
 
              for(auto child: adj[lastVis.first]){
                   if(!vis[child]){
                       q.push({child, lastVis.second + 1});
                       vis[child] = true;
                   }
              }
        }
 
 
       return lastVis;
 
 
}
int main(){
    
    int n;
    cin >> n;
 
    vector<int>adj[n+5];
 
    for(int i=1;i<n;i++){
       int a,b;
       cin >> a >> b;
       adj[a].push_back(b);
       adj[b].push_back(a);
        
    }
    
     pair<int,int>p = bfs(1, n,adj);
      
      pair<int,int>pp = bfs(p.first, n, adj);
 
   
 
     cout<< pp.second;
 
    return 0;
}

File: findMaxDistance.java
import java.util.*;

public class Main {
    
    static class Pair {
        int first;
        int second;
        
        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }
    
    static Pair bfs(int node, int n, ArrayList<Integer>[] adj) {
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(node, 0));
        
        boolean[] vis = new boolean[n + 1];
        vis[node] = true;
        
        Pair lastVis = null;
        
        while (!q.isEmpty()) {
            lastVis = q.poll();
            
            for (int child : adj[lastVis.first]) {
                if (!vis[child]) {
                    q.add(new Pair(child, lastVis.second + 1));
                    vis[child] = true;
                }
            }
        }
        
        return lastVis;
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        int n = scanner.nextInt();
        
        @SuppressWarnings("unchecked")
        ArrayList<Integer>[] adj = new ArrayList[n + 5];
        
        for (int i = 1; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for (int i = 1; i < n; i++) {
            int a = scanner.nextInt();
            int b = scanner.nextInt();
            adj[a].add(b);
            adj[b].add(a);
        }
        
        Pair p = bfs(1, n, adj);
        Pair pp = bfs(p.first, n, adj);
        
        System.out.println(pp.second);
        
        scanner.close();
    }
}
